src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs

```diff
@@ -0,0 +1,149 @@
+                       && invocation.Arguments[0].Value.Syntax.IsEquivalentTo(containsKeyArgument.Syntax)
+                       && invocation.Arguments[1].Value.Kind == OperationKind.Literal;
+            }
+
+            if (conditionalBody is IBlockOperation block)
+            {
+                dictionaryAddLocation = block.Operations.OfType<IExpressionStatementOperation>().FirstOrDefault(e => e.Operation is IInvocationOperation i && IsEligibleDictionaryAddInvocation(i))?.Operation.Syntax.GetLocation();
+            }
+            else if (conditionalBody is IExpressionStatementOperation { Operation: IInvocationOperation iOperation } && IsEligibleDictionaryAddInvocation(iOperation))
+            {
+                dictionaryAddLocation = iOperation.Syntax.GetLocation();
+            }
+
+            return dictionaryAddLocation is not null;
+        }
+
+        private static bool TryGetDictionaryTypeAndMembers(
+            Compilation compilation,
+            [NotNullWhen(true)] out INamedTypeSymbol? iDictionaryType,
+            [NotNullWhen(true)] out IMethodSymbol? containsKeySymbol,
+            [NotNullWhen(true)] out IMethodSymbol? addSymbol)
+        {
+            containsKeySymbol = null;
+            addSymbol = null;
+            iDictionaryType = WellKnownTypeProvider.GetOrCreate(compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericIDictionary2);
+            if (iDictionaryType is null)
+            {
+                return false;
+            }
+
+            containsKeySymbol = iDictionaryType.GetMembers().OfType<IMethodSymbol>().FirstOrDefault(m => m.Name == ContainsKey);
+            addSymbol = iDictionaryType.GetMembers().OfType<IMethodSymbol>().FirstOrDefault(m => m.Name == Add);
+
+            return containsKeySymbol is not null && addSymbol is not null;
+        }
+
+        private static bool IsDictionaryType(INamedTypeSymbol suspectedDictionaryType, ISymbol iDictionaryType)
+        {
+            // Either the type is the IDictionary or it is a type which (indirectly) implements it.
+            return suspectedDictionaryType.OriginalDefinition.Equals(iDictionaryType, SymbolEqualityComparer.Default)
+                   || suspectedDictionaryType.AllInterfaces.Any((@interface, dictionary) => @interface.OriginalDefinition.Equals(dictionary, SymbolEqualityComparer.Default), iDictionaryType);
+        }
+
+        private static bool DoesSignatureMatch(IMethodSymbol suspected, IMethodSymbol comparator)
+        {
+            return suspected.OriginalDefinition.ReturnType.Name == comparator.ReturnType.Name
+                   && suspected.Name == comparator.Name
+                   && suspected.Parameters.Length == comparator.Parameters.Length
+                   && suspected.Parameters.Zip(comparator.Parameters, (p1, p2) => p1.OriginalDefinition.Type.Name == p2.Type.Name).All(isParameterEqual => isParameterEqual);
+        }
```

Comment:
Why not use `suspected.Equals(comparator, SymbolEqualityComparer.Default)`?

Reply:
Because symbolically, `System.Collections.Generic.Dictionary<TKey, TValue>.ContainsKey(TKey)` and `System.Collections.Generic.IDictionary<TKey, TValue>.ContainsKey(TKey)` are not equal.

Reply:
@CollinAlpert There is an `IsImplementationOfInterfaceMember` extension method that I think could work for this scenario.

Reply:
Yeah, I tried substituting the body of the method with `return suspected.OriginalDefinition.IsImplementationOfInterfaceMember(comparator);`, but that also doesn't work.